### ISSUE TITLE
MNIST tutorial fails to train correctly when using Covalent

### ISSUE THREAD

**Issue by cjao** [AgnostiqHQ/covalent#646]
As others have already pointed out in [a different issue](https://github.com/AgnostiqHQ/covalent/issues/638). The MNIST tutorial fails to train the ML model when using Covalent. [Cell 14](https://covalent.readthedocs.io/en/latest/tutorials/machine_learning/mnist_classifier.html#Transform-and-run-workflow-with-Covalent) shows a training accuracy of approximately 10%, the same accuracy as from random guessing. 

The reason is that we're "using Covalent wrong". Users must remember that electrons cannot manipulate external data structures passed by reference. Since the electron can potentially run in a separate process or even on a separate computer from the Covalent server,  any references to objects returned from another electron (which are stored in the Covalent server's memory) will surely break. Consequently, **Covalent must make deep copies of all inputs to each electron***.

Here's an excerpt from **Cell 10** of the tutorial:
```
    optimizer = get_optimizer(model=model, learning_rate=learning_rate, momentum=momentum)
    model, optimizer = train_model(
        train_dataloader=train_dataloader,
        train_losses=train_losses,
        train_counter=train_counter,
        log_interval=log_interval,
        model=model,
        optimizer=optimizer,
        loss_fn=loss_fn,
        epochs=epochs,
    )
```
`get_optimizer` and `train_model` are separate electrons, defined in Cells 6 and 9 respectively.  Here's `get_optimizer`:
```
def get_optimizer(
    model: NeuralNetwork, learning_rate: float, momentum: float
) -> torch.optim.Optimizer:
    """Get Stochastic Gradient Descent optimizer."""

    return torch.optim.SGD(model.parameters(), learning_rate, momentum)

```
Now let's take a look at `train_model`:
```
def train_model(
    train_dataloader: torch.utils.data.dataloader.DataLoader,
    train_losses: list,
    train_counter: int,
    log_interval: int,
    model: NeuralNetwork,

    loss_fn: callable,
    epochs: int,
    results_dir: str = "~/data/mnist/results/",
) -> Tuple[NeuralNetwork,]:
...
    for epoch in range(1, epochs + 1):
        print(f"Epoch {epoch}\n-------------------------------")
        model, optimizer = train_over_one_epoch(
            dataloader=train_dataloader,
            model=model,
            optimizer=optimizer,
            train_losses=train_losses,
            train_counter=train_counter,
            log_interval=log_interval,
            epoch=epoch,
            loss_fn=loss_fn,
        )
...
```
When this electron is executed, the `optimizer` argument refers to a **different** Optimizer instance from what was returned by `get_optimizer` in Cell 10. Indeed if we want to run this electron using the Slurm executor or in a separate worker process using Dask, we must make deep copies of all inputs; a remote server has no way to manipulate the Optimizer instance living on the Covalent server. Similarly, `model` is an independent copy of the `model` defined in the original workflow. 

**These two objects are not connected**. When `train_model` calls `optimizer.step()` (see `train_over_one_epoch`), the `optimizer` object modifies its own copy of the model parameters while the parameters in `model` remain unchanged.

One remedy is to eliminate the `get_optimizer` electron and instead instantiate the Optimizer inside `train_model` using the model passed to `train_model`, something like
```
@ct.electron
def train_model(
    train_dataloader: torch.utils.data.dataloader.DataLoader,
    train_losses: list,
    train_counter: int,
    log_interval: int,
    model: NeuralNetwork,
    learning_rate: float,
    momentum: float,
    optimizer: torch.optim.Optimizer,
    loss_fn: callable,
    epochs: int,
    results_dir: str = "~/data/mnist/results/",
) -> Tuple[NeuralNetwork,]:
    """Train neural network model."""

   optimizer = torch.optim.SGD(model.parameters(), learning_rate, momentum)
    for epoch in range(1, epochs + 1):
        print(f"Epoch {epoch}\n-------------------------------")
        model, optimizer = train_over_one_epoch(
            dataloader=train_dataloader,
            model=model,
            optimizer=optimizer,
            train_losses=train_losses,
            train_counter=train_counter,
            log_interval=log_interval,
            epoch=epoch,
            loss_fn=loss_fn,
        )
    ...
    return model, optimizer

```

*This tutorial worked in older versions of Covalent (v32.3 and before) because the local executor that was used by default didn't make deep copies of inputs to electrons. I would expect the tutorial to fail if one had tried a remote executor for `train_model`. [This commit](https://github.com/AgnostiqHQ/covalent/commit/1a4fa1e99d186629cc439fbf30792ce7a9ffa801#diff-a99ca6e887d5a6439efcda540e9e2661cdafa9e3a92109f2ac9c390299f794fbR197) introduced (effectively) a deep copy for all executors, including local. And nowadays, we default to a multiprocessing-based Dask executor which would require copies of inputs even without that commit.

**Comment by venkatBala** [comment]
@cjao this was fixed if I recall correctly?

**Comment by cjao** [comment]
Yes, I think this can be closed now.
